ocaml/xenopsd/xc/service.ml

```diff
@@ -256,31 +258,56 @@ module DaemonMgmt (D : DAEMONPIDPATH) = struct
 
   let name = D.name
 
-  let pid ~xs domid =
+  (* Look up the commandline args for the process pid; *)
+  (* Check that they include expected args. *)
+  let is_cmdline_valid pid expected_args =
     try
-      match D.pid_location with
-      | (File file | Path_of {file; _}) when Sys.file_exists (file domid) ->
-          let path = file domid in
-          let ( let* ) = Option.bind in
-          let* pid = Unixext.pidfile_read path in
-          Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
```

Comment:
Could you please explain why this check doesn't work?
The expectation here is that if the target process has terminated, this pidfile should have been unlocked, so `None` is supposed to be returned. But the issue is `Some pid` was returned.

Reply:
Sure, I commented in the ticket, OK, let me also explain here. 

We can reproduce the problem this way:

1. setup a host with a rhel guest running.
2. kill the qemu for that guest in host console.
3. "Force Shut Down" the rhel guest in XC
4. xapi still try to kill the process with the log: "qemu-dm: stopping qemu-dm with SIGTERM"

Below is the details about how it is reproduced:

when the VM is "Force Shut Down" in XC, xenopsd-xc received 2 stop requests
1. for the first one, it did not return the pid as it found the pid invalid because we succeeded taking the lock: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L269-L271, but it removed the pidfile: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L380-L381
2. for the second stop request, when checking for pid, it did not match the `File` case as the pidfile doesn't exist: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L262, it matches the `Xenstore` case and returned the pid stored in Xenstore: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L279-L280, and then it will continue to kill the process with pid as process id: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/service.ml#L391

Reply:
I would suggest to add logging to see exactly which case is responsible for returning the PID.

Reply:
Log added: https://github.com/xapi-project/xen-api/commit/40016d998d3cc402cf390fa4c7220833d2edfa0b